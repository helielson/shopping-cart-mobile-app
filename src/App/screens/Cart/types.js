/**
 * Types definition used on Cart screen
 *
 * @format
 * @flow
 */

// PropTypes for Cart screen Component
export type CartPropTypes = {
  items: Array<{}>,
  itemsCount: number,
  total: number,
};
