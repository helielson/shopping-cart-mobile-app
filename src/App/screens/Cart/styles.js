import { Platform } from 'react-native';
import colors from '../../config/colors';

export default {
  container: {
    backgroundColor: colors.background,
    flex: 1,
  },
  footBar: {
    backgroundColor: colors.backgroundLight,
    flexDirection: 'row',
    alignItems: 'flex-end',
    alignSelf: 'flex-end',
    paddingHorizontal: 20,
    paddingVertical: 15,
    paddingBottom: Platform.OS === 'ios' ? 50 : 20,
  },
  previewContainer: {
    flex: 1.5,
    flexDirection: 'column',
  },
  itemsCount: {
    color: colors.ternary,
  },
  totalContainer: {
    fontSize: 17,
    color: '#444',
    fontWeight: 'bold',
  },
  total: {
    fontSize: 17,
    color: colors.primary,
    fontWeight: 'bold',
  },
  emptyContainer: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  emptyText: {
    fontSize: 20,
    color: colors.primary,
    fontWeight: 'bold',
  },
};
