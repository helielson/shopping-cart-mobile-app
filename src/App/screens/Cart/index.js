/**
 * @format
 * @flow
 */
import React from 'react';
import { View, FlatList, Text } from 'react-native';
import { connect } from 'react-redux';
import type { CartPropTypes } from './types';
import CartItem from '../../components/CartItem';
import Button from '../../components/Button';
import formatPrice from '../../util';
import colors from '../../config/colors';
import styles from './styles';

const keyExtractor = (item: {
  id: string,
  description: string,
  price: number,
  qtd: number,
  available: number,
  picture: string,
  title: string,
}): string => item.id;

const renderFooter = (total: number, itemsCount: number) => (
  <View style={styles.footBar}>
    <View style={styles.previewContainer}>
      <Text style={styles.itemsCount}>
        ({itemsCount} {itemsCount === 1 ? 'item' : 'items'})
      </Text>
      <Text style={styles.totalContainer}>
        Total: <Text style={styles.total}>{formatPrice(total)}</Text>
      </Text>
    </View>
    <Button text="Checkout" disabled color={colors.backgroundLight} onPress={() => {}} />
  </View>
);

const renderEmpty = () => (
  <View style={styles.emptyContainer}>
    <Text style={styles.emptyText}>Your Cart is Empty!</Text>
  </View>
);

/**
 * Component to render the list of items already
 * on user cart
 *
 * @param {CartPropTypes} {
 *   items,
 *   itemsCount,
 *   total
 * }
 * @returns <Cart />
 */
const Cart = ({ items, itemsCount, total }: CartPropTypes) => (
  <View style={styles.container}>
    {items.length ? (
      <FlatList
        data={items}
        keyExtractor={keyExtractor}
        renderItem={({ item }) => <CartItem item={item} key={item.id} />}
      />
    ) : (
      renderEmpty()
    )}
    {renderFooter(total, itemsCount)}
  </View>
);

// Map Redux States to Component Props
const mapStateToProps = state => ({
  items: Object.values(state.cart.items),
  itemsCount: state.cart.itens_count,
  total: state.cart.total_price,
});

// Export Stateless Component not connected to redux store (To be used on Tests)
export { Cart };
// Connect component with store and export it as default
export default connect(mapStateToProps)(Cart);
