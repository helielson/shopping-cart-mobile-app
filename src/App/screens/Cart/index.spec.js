/** @format */

import React from 'react';
import { shallow } from 'enzyme';

import Cart from '.';

describe('<Cart />', () => {
  test('should render properly', () => {
    const wrapper = shallow(<Cart />);
    expect(wrapper).toMatchSnapshot();
  });
});
