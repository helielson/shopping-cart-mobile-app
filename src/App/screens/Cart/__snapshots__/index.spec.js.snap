// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cart /> should render properly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "rgba(225, 226, 225, 1.00)",
      "flex": 1,
    }
  }
>
  <FlatList
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReachedThreshold={2}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</View>
`;
